src/parser.rs

```diff
@@ -204,6 +231,60 @@ impl<'a> Parser<'a> {
         }
     }
 
+    fn parse_unary_plus_expression(&mut self) -> Option<UnaryPlusExpression> {
+        assert!(self.token.kind() == TokenKind::Plus);
+        self.read_token(); // consume `+`
+        match self.token.kind()
+        {
+            TokenKind::Integer => {
+                if let Some(integer) = self.parse_integer_expression() {
+                    return Some(UnaryPlusExpression::IntegerExpression(integer));
+                }
+                None
+            }
+            TokenKind::FloatingPoint => {
+                if let Some(floating_point) = self.parse_floating_point_expression() {
+                    return Some(UnaryPlusExpression::FloatingPointExpression(floating_point));
+                }
+                None
+            }
+            _ => {
+                self.errors.push(format!(
+                        "Parse error when parsing unary plus expression {:?}",
+                        self.token
+                        ));
+                None
+            }
+        }
+    }
+
+    fn parse_unary_minus_expression(&mut self) -> Option<UnaryMinusExpression> {
```

Comment:
I'm a little unhappy with the code duplication here. I did try to use a generic function and parameterise on the `UnaryPlusExpression` or `UnaryMinusExpression`, but it didn't like that, and I don't know enough about rust generics yet to understand why (beyond that an `enum` is not a type, apparently). I'll come back to this when I've gained this understanding.